src/attributes/diagnostics.md

```diff
@@ -252,10 +252,125 @@ When used on a function in a trait implementation, the attribute does nothing.
 > let _ = five();
 > ```
 
+# The `non_exhaustive` attribute
+
+The *`non_exhaustive` attribute* is used to indicate that a type will have
+more fields or variants added in the future. It can be applied to
+[`struct`s][struct], [`enum`s][enum] and `enum` variants.
+
+The `non_exhaustive` attribute uses the [_MetaWord_] syntax and thus does not
+take any inputs.
+
+Within the defining crate, types annotated with `non_exhaustive` behave exactly
+the same as if the type were not annotated with `non_exhaustive`:
+
+```rust,ignore
+#[non_exhaustive]
+pub struct Config {
+    pub window_width: u16,
+    pub window_height: u16,
+}
+
+#[non_exhaustive]
+pub enum Error {
+    Message(String),
+    Other,
+}
+
+pub enum Message {
+    #[non_exhaustive] Send { from: u32, to: u32, contents: String },
+    #[non_exhaustive] Reaction(u32),
+    #[non_exhaustive] Quit,
+}
+
+// Non-exhaustive structs can be constructed as normal within the defining crate.
+let config = Config { window_width: 640, window_height: 480 };
+
+// Non-exhaustive structs can be matched on exhaustively within the defining crate.
+if let Ok(Config { window_width, window_height }) = config {
+    // ...
+}
+
+// Non-exhaustive enums can be matched on exhaustively within the defining crate.
+match error {
+    Error::Message(ref s) => { },
+    Error::Other => { },
+}
+
+match message {
+    // Non-exhaustive variants can be matched on exhaustively within the defining crate.
+    Message::Send { from, to, contents } => { },
+    Message::Reaction(type) => { },
+    Message::Quit => { },
+}
+```
+
+In downstream crates, types annotated with `non_exhaustive` have limitations that
+preserve backwards compatibility when new fields or variants are added.
+
+Non-exhaustive types cannot be constructed in downstream crates:
+
+```rust,ignore
+// `Config`, `Error` and `Message` are types defined in an upstream crate that have been
+// annotated as `#[non_exhaustive]`.
+use upstream::{Config, Error, Message};
+
+// Cannot construct an instance of `Config`, if new fields were added in
+// a new version of `upstream` then this would fail to compile, so it is
+// disallowed.
+let config = Config { window_width: 640, window_height: 480 };
+
+// Can construct an instance of `Error`, new variants being introduced would
+// not result in this failing to compile.
+let error = Error::Message("foo".to_string());
+
+// Cannot construct an instance of `Message::Send` or `Message::Reaction`,
+// if new fields were added in a new version of `upstream` then this would
+// fail to compile, so it is disallowed.
+let message = Message::Send { from: 0, to: 1, contents: "foo".to_string(), };
+let message = Message::Reaction(0);
+
+// Cannot construct an instance of `Message::Quit`, if this were converted to
+// a tuple-variant `upstream` then this would fail to compile.
+let message = Message::Quit;
+```
+
+Non-exhaustive types cannot be used in a [`match`] or [`if let`] expression without a
```

Comment:
And `while let`, `for`, and `let` as well as function parameters?
Also, talking about adding a wildcard arm to `if let` seems strange; it already has one implicitly.

I think we can describe things as (modulo rewording -- and do check my logic):

- When pattern matching on a non_exhaustive `enum`, matching on a variant does not contribute towards exhaustiveness of the arms. (`_ =>` falls out from this)
- When pattern matching on a non_exhaustive variant (struct, enum variant), struct patterns must be used and  `..` must be included in in them.

(Ideally we reference entities from the pattern grammar.)

Reply:
I've updated the wording based on what you provided here, and added some references to the patterns section.